doc/reference/std/misc/atom.md

```diff
@@ -0,0 +1,339 @@
+which is equivalent to the function `true` that accepts every value.
+The current value is grandfathered and not validated;
+the validator is not validated for whether it accepts the current value.
+The function is not guaranteed to return any specific value.
+Similar to Clojure `set-validator!`.
+
+::: tip Examples:
+``` scheme
+> (def a (atom 42 validator: number?))
+> (atom-reset! a 1)
+1
+> (atom-validator-set! a string?)
+> (atom-reset! a "foo")
+"foo"
+> (atom-reset! a 2)
+*** ERROR IN "misc/atom.ss"@48.29 [BadArgument]: bad argument; expected valid atom value
+--- irritants: update-atom! 1
+```
+:::
+
+## atom-add-watch!
+
+``` scheme
+(atom-add-watch! atom key function) -> void
+```
+
+Adds a watch function to an atom.
+The watch function must be a function of 4 args:
+a key, the atom, its old-state, its new-state.
+Whenever the atom's state is changed,
+after the new value has been validated,
+each of the registered watches will have their functions called,
+in an unspecified order, while the atom lock is still held
+(so no changes to the atom are allowed in the watches themselves,
+though they may send asynchronous messages or spawn threads that
+will cause a cascade of further changes).
+Keys must be unique per atom, and can be used to remove the watch with
+`atom-remove-watch!`, but are otherwise considered opaque
+by the watch mechanism.
+
+It is unspecified whether the state has been modified yet,
+so use old-state and new-state rather than atom-deref.
+Note also that watch functions may be called from multiple threads
+simultaneously. (In current implementation, the state will already
+have been modified and no threads will be spawned.)
+Similar to Clojure's `add-watch` function.
+
+::: tip Examples:
+``` scheme
+> (import :std/misc/atom :std/misc/list)
+> (def transitions [])
+> (def a (atom 1))
+> (atom-add-watch! a 'record-transitions
+    (lambda (_k _a old new) (push! [old new] transitions)))
+> (atom-swap! a (cut * 2 <>))
+2
+> (atom-swap! a (cut * 2 <>))
+4
+> transitions
+((2 4) (1 2))
+> (atom-remove-watch! a 'record-transitions)
+> (atom-swap! a (cut * 2 <>))
+8
+> transitions
+((2 4) (1 2))
+```
+:::
+
+## atom-remove-watch!
+
+``` scheme
+(atom-remove-watch! atom key) -> void
+```
+
+Removes a watch function to an atom.
+Similar to Clojure's `remove-watch` function.
+See example for `atom-add-watch!` above.
+
+## atom-increment!
+
+``` scheme
+(atom-increment! atom [increment]) -> number
+```
+
+Increments the atom value by the specified increment, or by 1 if unspecified.
+Return the new value. Common special case for a numeric (often integer) atom.
+
+::: tip Examples:
+``` scheme
+> (import :std/misc/atom)
+> (def a (atom -1))
+> (atom-increment! a)
+0
+> (atom-increment! a)
+1
+> (atom-increment! a)
+2
+```
+:::
+
+## atomic-counter
+
+``` scheme
+(atomic-counter [initial-value]) -> function
+```
+
+Creates a function that maintains an atomic counter
+that starts with the given initial-value (which defaults to -1),
```

Comment:
Creates a function that maintains an atomic counter that starts with the given `initial-value` (which defaults to -1). The function takes an `increment` as an optional argument, which defaults to 1. The function increments the atomic value, returning the new value.